Add controller support for Pioneer DDJ-400

WarkerAnhaltRanger opened the pull request:
This is a initial Support for the Pioneer DDJ-400
Testers are welcome!
Working:
                * Mixer Section (Faders, EQ, Filter, Gain, Cue)
                * Browsing and loading + Waveform zoom (shift)
                * Jogwheels, Scratching, Bending
                * cycle Temporange
                * Beat Sync
                * Beat Loop Mode
                * Sampler Mode
                * Beatjump mode
            
Testing:
                * Keyboard Mode (check pitch value)
                * Keyshift Mode (check pitch value
                * Hot Cue Mode (including loops)
                * Loop Section: Loop in / Out + Adjust, Call, Double, Half
                * Effect Section (Beat FX left + Right - select the Effect Slot of FX3 (not Effect BPM))
                * Output (lights)

Partially:
                * PAD FX (only slots A-H, Q-P)
                * Effect Section (without Beat FX left + Right - no equivalent function found)

Not working/implemented:
                * Channel & Crossfader Start

WarkerAnhaltRanger committed:
Version 0.2:  * added Beatjump Beatcount manipulation increase (pad 8 + shift), decrease (pad 7 +shift)
                * fixed a bug in beatloop with 1/4 and 1/2 Beat loops
                * fixed a bug in BEAT SYNC + Shift (cycle tempo range)
                * added support for saving and playing loops on Hot Cue Pads
                * added support for loop in + out adjust (while looping press and hold in or out + jog wheel to adjust the loop poit position)
                * fixed a bug in Tempo Range switch (Beat Sync + shift)

WarkerAnhaltRanger committed:
changed description to match the convention

WarkerAnhaltRanger committed:
renaming  issues fixed

WarkerAnhaltRanger committed:
added Effect Section funtions

WarkerAnhaltRanger committed:
initial Output / LED support

WarkerAnhaltRanger committed:
support for some LEDs and cue loop left and right

WarkerAnhaltRanger committed:
hotcue and beatloop LEDs added

WarkerAnhaltRanger committed:
reloop/exit LED on if loop_start_position > 0 | loop_enabled > 0

WarkerAnhaltRanger committed:
Sampler PAD LED on if track loaded

WarkerAnhaltRanger committed:
keyboard mode LEDs implemented (untested)

WarkerAnhaltRanger committed:
Pad FX and Keyshift LEDs

WarkerAnhaltRanger commented:
* Forum: https://mixxx.org/forums/viewtopic.php?f=7&t=12113
* Wiki: https://www.mixxx.org/wiki/doku.php/pioneer_ddj-400

WarkerAnhaltRanger committed:
added wiki and forumlink

WarkerAnhaltRanger committed:
typo fixed

WarkerAnhaltRanger committed:
copy paste errors fixed

WarkerAnhaltRanger commented:
- [ ] implement blinking pad when playing sample or fx
- [ ] implement blinking Loop in and out when loop is activated

WarkerAnhaltRanger committed:
prefixed unused function params with underscore

WarkerAnhaltRanger committed:
removed unused func params to satisfact codefactor

WarkerAnhaltRanger committed:
removed unsused vars

WarkerAnhaltRanger commented:
could someone review this?

uklotzde commented:
Thank you for providing this mapping! We need the permission to distribute your contributions by signing the [Mixxx Contributor Agreement](https://docs.google.com/forms/d/e/1FAIpQLScC9QG327sjLL0eWftmfGUasxFFLxg6LCXJ2xHDYRzFIRqyiw/viewform?formkey=dEpYN2NkVEFnWWQzbkFfM0ZYYUZ5X2c6MQ)

I have some initial technical remarks:
- Please add a detailed description of the mapping to the wiki, example: https://www.mixxx.org/wiki/doku.php/roland_dj-505
- The use of whitespace in the JS file is very inconsistent. Please consider reformatting it to improve readability.
- I would recommend to run JSHint on the JS file. This should result in some recommendations like using ===/!== instead of ==/!=.

And please be patient if you don't get feedback within a few hours. We are all volunteers that spend their rare spare time for this project.

WarkerAnhaltRanger commented:
Thank you for your Feedback.
Signed the agreement.
Going to document mapping in the wiki soon. Basically I implemented the same functions as in the User manual for Rekordbox. 
JSHint did not state any problems besides that hotcuePad in keyboardModeEnabledOutput was already defined. Fixed that.
I'll check the format.

WarkerAnhaltRanger committed:
fixed double assignment

WarkerAnhaltRanger committed:
some formatting

nschloe commented:
I've tried this PR with my DDJ-400 and found it _mostly_ working. At first glance, things not working:

 * Track selector scrolls fine, but doesn't load track on click.
 * Pressing  any of `HOT CUE`, `BEAT LOOP`, `BEST JUMP`, `SAMPLER` results in the error message
   `Uncaught exception at line 1 in passed code: TypeError: Result of expression '(PioneerDDJ400.modeChange)' [undefined] is not a function.`

uklotzde commented:
It's ok to add missing features later, but at least the undefined access errors should be fixed.

uklotzde commented:
@nschloe Thanks for testing!

WarkerAnhaltRanger committed:
removed reference to removed Function modeChange

WarkerAnhaltRanger commented:
> I've tried this PR with my DDJ-400 and found it _mostly_ working. At first glance, things not working:

Thank you for testing!

> 
>     * Track selector scrolls fine, but doesn't load track on click.
> 
The press on the rotary selector cycles the focus between the Library Folders and the Tracks. (same as Rekordbox). You load the Track using the "LOAD" button 

>     * Pressing  any of `HOT CUE`, `BEAT LOOP`, `BEST JUMP`, `SAMPLER` results in the error message
>       `Uncaught exception at line 1 in passed code: TypeError: Result of expression '(PioneerDDJ400.modeChange)' [undefined] is not a function.`
I accidently forgot to remove the reference for the removed modeChange function in the XML File.
I updated the XML.

nschloe commented:
> The press on the rotary selector cycles the focus between the Library Folders and the Tracks. (same as Rekordbox). You load the Track using the "LOAD" button

Perfect, that works.

> I updated the XML.

Confirmed fixed.

WarkerAnhaltRanger committed:
removed local gitignore

WarkerAnhaltRanger commented:
I updated the Wikipage.
The current implementation is covered.
Also I made a Mistake with the .gitignore file (and I don't know how to revert this). The .gitignore File should not be merged.

nschloe commented:
> The .gitignore File should not be merged.

Simply `git revert` the respective commit.

nschloe commented:
Btw, I've been using this PR succussfully with my DDJ-400 for a while now. How can I help getting this merged?

jsBrian commented:
I've been using this branch with my DDJ-400 on Ubuntu 18.04.  Works for me as well.

Commit:
feat: added blinking lights to sampler

WarkerAnhaltRanger committed:
Merge pull request #1 from jsBrian/feature/blinking-sampler

feat: added blinking lights to sampler (thx to jsbrian)

Commit:
feat: better fx handling

Commit:
feat(controller): DDJ-400 beat fx channel selector improvement

WarkerAnhaltRanger committed:
Merge pull request #2 from jsBrian/feature/better-fx-handling

feat: better fx handling (thx jsBrian)

WarkerAnhaltRanger commented:
Is there something missing or wrong with the code? How can I get this merged into mixxx?

unlimited67g commented:
Hello all,

first thank you for your work.

I'm quite a noob.
Trying to make my DD 400 work with linux mint 19.2 cinnamon and Mixxx 2.2 . 

I'm putting the .xml and .js files in the controllers folder but get this issue :

`
There was a problem parsing the XML file /usr/share/mixxx/controllers/Pioneer-DDJ-400.midi.xml.

unexpected character at line 44, column 89`

Any help welcome !

thanks again ;)

uklotzde commented:
@WarkerAnhaltRanger Please _interactively rebase_ your branch and remove all unrelated changes that affect files other than the added .xml and the .js file.

uklotzde commented:
@unlimited67g Please verify that you downloaded the actual contents of the files (_Raw_) and not some HTML view from Github. I don't spot any errors on line 44 in the .xml file.

Be-ing commented:
Sorry for the delay. I'll review this today or tomorrow.

Be-ing commented:
Unless you have mapped new features for Mixxx 2.3, please change the target branch of the pull request to 2.2 and rebase on the 2.2 branch. That way this can be included in Mixxx 2.2.3.

Be-ing commented on res/controllers/Pioneer-DDJ-400-script.js:
```diff
@@ -0,0 +1,718 @@
+PioneerDDJ400.numFxSlots = 3;
+
+Object.defineProperty(PioneerDDJ400, 'selectedFxSlot', {
+    get: function() {
```
This works with the ancient QtScript interpreter? I remember trying this years ago and couldn't get it to work.

Be-ing commented on res/controllers/Pioneer-DDJ-400-script.js:
```diff
@@ -0,0 +1,718 @@
+});
+
+PioneerDDJ400.padFxBelowPressed = ignoreRelease(function(channel, control, value, status, group) {
+    const groupAbove = group.replace(/\[EffectRack1_EffectUnit(\d+)_Effect(\d+)]/, function(all, unit, effect) {
```
`const` is not actually supported by the interpreter; that it happens to work is only a coincidence. Please change all `const`s to `var` to avoid confusion.

Be-ing commented on res/controllers/Pioneer-DDJ-400-script.js:
```diff
@@ -0,0 +1,718 @@
+
+PioneerDDJ400.padFxBelowPressed = ignoreRelease(function(channel, control, value, status, group) {
+    const groupAbove = group.replace(/\[EffectRack1_EffectUnit(\d+)_Effect(\d+)]/, function(all, unit, effect) {
+        const effectAbove = parseInt(effect, 10) - 4;
```
I don't think the `10` parameter for `parseInt` is needed.

Be-ing commented on res/controllers/Pioneer-DDJ-400-script.js:
```diff
@@ -0,0 +1,718 @@
+};
+
+PioneerDDJ400.cueLoopCallLeft = function(_channel, _control, value, _status, group){
+    if(value == 0) return; // ignore release
```
Please do not use one line `if` statements.

Be-ing commented on res/controllers/Pioneer-DDJ-400-script.js:
```diff
@@ -0,0 +1,718 @@
+        // loop halve
+        engine.setValue(group, 'loop_scale', 0.5);
+    }
+    else{
```
Move this to the previous line: `} else {`

Be-ing commented on res/controllers/Pioneer-DDJ-400-script.js:
```diff
@@ -0,0 +1,718 @@
+    if(value == 0) return; // ignore release
+
+    const loop_on = engine.getValue(group, 'loop_enabled');
+    if (loop_on){
```
space after `)`:
`if (loop_on) {`

Be-ing commented on res/controllers/Pioneer-DDJ-400-script.js:
```diff
@@ -0,0 +1,718 @@
+            }
+        }
+        //engine.setValue(group, 'loop_in_goto', 1);
+        engine.setValue(group, 'playposition', newPosition);
```
I do not understand the purpose of this. This controller has hotcue buttons, so what does this add?

Be-ing commented on res/controllers/Pioneer-DDJ-400-script.js:
```diff
@@ -0,0 +1,718 @@
+
+    // loop_in / out adjust
+    const loopEnabled = engine.getValue(group, 'loop_enabled');
+    if(loopEnabled > 0){
```
spaces around parentheses: `if (loopEnabled > 0) {`

Be-ing commented on res/controllers/Pioneer-DDJ-400-script.js:
```diff
@@ -0,0 +1,718 @@
+        return;
+    }
+
+    // save playing loop on hotcue pad
```
I think this hack should be removed. What is the point of saving a loop if it is lost when Mixxx shuts down? We are currently working on adding this feature to Mixxx in #2194. If you would like to help implement and/or test that feature, that would be great!

Be-ing commented on res/controllers/Pioneer-DDJ-400-script.js:
```diff
@@ -0,0 +1,718 @@
+
+Object.defineProperty(PioneerDDJ400, 'selectedFxSlot', {
+    get: function() {
+        return engine.getValue('[EffectRack1_EffectUnit3]', 'focused_effect');
```
Why does this use EffectUnit 3? This requires the user to show all 4 effect units on screen even though the controller can only manipulate one of them. Wouldn't it make more sense to map the controller to EffectUnit 1?

Be-ing commented on res/controllers/Pioneer-DDJ-400.midi.xml:
```diff
@@ -0,0 +1,4688 @@
+            <control>
+                <description>RELOOP/EXIT +SHIFT (DECK1) - press - Active loop on/off</description>
+                <group>[Channel1]</group>
+                <key>reloop_toggle</key> <!-- TODO Check if correct -->
```
I suggest mapping this to `reloop_andstop`

Be-ing reviewed (changes requested):
Some preliminary comments; I haven't read through the whole wiki or script yet. There are lots of little style issues. Please review the [JS coding conventions](https://mixxx.org/wiki/doku.php/contributing_mappings#coding_conventions_for_javascript)

unlimited67g commented:
> @unlimited67g Please verify that you downloaded the actual contents of the files (_Raw_) and not some HTML view from Github. I don't spot any errors on line 44 in the .xml file.

Hello,

thank you i did that and no more of this error !

But when i plug the controller and the open mixxx and then enable the midi controller in the preferences, i can't control anything with the controller...
Am i missing something (again..) ?

Thanks

WarkerAnhaltRanger commented:
I'm sorry. I'm afraid I can't finish this pull request since my RL took over.
Maybe someone else could finish this?

Be-ing commented:
That's okay, thank you for communicating that. @nschloe or @jsBrian would you like to continue with this? You can start by making a branch on your fork from this branch and making a new pull request.

nschloe commented:
See #2312.
